### PR TITLE
Fine grained data collection

### PR DESCRIPTION
**Pull request by Zsailer**
This replaces #46. The PR was originally directed at handling personal data. Now, it's a bit more general.

Here, I've changed the `allowed_schemas` trait to explicitly list all event properties that should be recorded based on their name or category. This can be used to filter-out personal data.

Summary of the changes made:
* `allowed_schemas` can now be a dictionary describing *what* data to keep when recording an event (see example below).
* schemas require a `categories` field under each property. This field can be used to group properties.
* `unrestricted` is a special category that means the named property will always be recorded when that event is recorded.

## How it works

`allowed_schemas` is now a nested dictionary where the keys are the schema IDs, and the values are dictionaries with two optional fields, `properties` and `categories`, describing the properties in the event that should be recorded.

As an example, 
```python
allowed_schemas = {
    "uri.to.schema": {
        "allowed_properties": ["name", "email"],
        "allowed_categories": ["personal-information"]
    }
}


c.EventLog.allowed_schemas=allowed_schemas
```
When the `uri.to.schema` event is recorded, it will include the properties, `email` and `name`, **and** any other properties with the category label `personal-information`. All other properties will not be recorded.

## Personal data

While this PR doesn't specifically address personal data, it provides a mechanism to filter out personal data. For example, properties can be labeled with categories like `hippa`, `gdpr`, `pii`, etc. to make filtering under different regimes possible.

### CODE REVIEW

File: jupyter_telemetry/eventlog.py [hunk 1]
```diff
@@ -112,21 +108,66 @@ def register_schema(self, schema):
...
+            try:
+                cats = attrs['categories']
+                # Categories must be a list.
+                if not isinstance(cats, list):
+                    raise ValueError(
+                        'The "categories" field in a registered schemas must be a list.'
+                    )
+
+                # Unrestricted is a special case and must be listed alone:
+                if 'unrestricted' in cats and len(cats) > 1:
```
Comment (Zsailer) [inline jupyter_telemetry/eventlog.py:134]: Relax this constraint. Schema authors should be able to list other categories with `unrestricted`.

File: jupyter_telemetry/traits.py [hunk 2]
```diff
@@ -40,3 +40,38 @@ def validate(self, obj, value):
...
+                           schema_name=schema_name,
+                           unknown_keys=",".join(unknown_keys)
+                        )
+                    )
+            validated_val = val
+        # If the type is a list (for backwards compatibility).
+        elif type(val) is list:
+            validated_val = {}
+            for schema_name in val:
+                validated_val[schema_name] = {"categories": []}
```
Comment (Zsailer) [inline jupyter_telemetry/traits.py:74]:
```suggestion
                validated_val[schema_name] = {}
```

File: docs/pages/application.rst [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+EventLog has two configurable traits:
+
+    - ``handlers``: a list of Python's logging handlers that handle the recording of incoming events.
+    - ``allowed_schemas``: a dictionary of options for each schema describing what data should be collected.
+
+Next, you'll need to register event schemas for your application. You can register schemas using the ``register_schema_file`` (JSON or YAML format) or ``register_schema`` methods.
+
+
+Once your have an instance of ``EventLog`` and your registered schemas, you can use the ``record`` method to log that event.
```
Comment (blink1073) [inline docs/pages/application.rst:34]: record_event

File: docs/pages/application.rst [hunk 4]
```diff
@@ -0,0 +1,44 @@
+.. _adding-telemetry:
+
+Adding telemetry to an application
+==================================
+
+Jupyter Telemetry enables you to log events from your running application. (It's designed to work best with traitlet's applications for simple configuration.) To use telemetry, begin by creating an instance of ``EventLog``:
```
Comment (blink1073) [inline docs/pages/application.rst:6]: application objects ?

File: docs/pages/schemas.rst [hunk 5]
```diff
@@ -31,13 +30,21 @@ Here is a minimal example of a valid JSON schema for an event.
...
-    - name
+    - thing
+    - user
+
 
+.. _JSON schema: https://json-schema.org/
 
-.. _JSON schema: https://json-schema.org/
\ No newline at end of file
+It is important that every schema explicity state whether it collects personal-data or not. See the :ref:`Personal Data` page for more information.
```
Comment (blink1073) [inline docs/pages/schemas.rst:50]: Does the Personal Data page exist?

File: tests/test_register_schema.py [hunk 6]
```diff
@@ -53,9 +53,11 @@ def test_reserved_properties():
         el.register_schema({
             '$id': 'test/test',
             'version': 1,
+            'personal-data': False,
```
Comment (blink1073) [inline tests/test_register_schema.py:56]: This field isn't in the base schema, correct?

File: docs/pages/application.rst (hunk 3 above)
Comment (Zsailer) [reply docs/pages/application.rst]: Thanks! Fixed.

File: docs/pages/application.rst (hunk 4 above)
Comment (Zsailer) [reply docs/pages/application.rst]: Updated!

File: tests/test_register_schema.py (hunk 6 above)
Comment (Zsailer) [reply tests/test_register_schema.py]: Oops, this was supposed to be removed.

**Comment by yuvipanda** [comment]
Thanks a lot for this, @Zsailer and @blink1073. 

I've one suggestion - when specifying the schema, can we call the properties `recorded_properties` and `recorded_categories`? That makes it much clearer what they are than now.

For categories, I'm worried that different schemas will use different words to refer to the same concepts. I don't think we should enforce that with software, but by providing useful guidance & some well known category names. For that, I propose:

1. Category names are also URIs, similar to schemas. They don't *have* to be, but by convention they are.
2. We define (in docs?) some well-known categories that can be used by schemas, and use them in our project jupyter schemas. Examples would be: `category.jupyter.org/user-identifier`, `category.jupyter.org/unrestricted`, `category.jupyter.org/action-timestamp`, etc. 

    This would mean we change 'unrestricted' to something like 'category.jupyter.org/unrestricted' (or similar? it's too long...), and document some more. The documentation doesn't have to happen in this PR though.

**Comment by yuvipanda** [comment]
From the admin perspective, this now requires the admin to:

1. Read the schema to understand what fields there are, and what categories
2. Figure out the fields they actually want in their emitted events, with a combination of categories and field names. This is an 'OR' list - fields will be included based on two of their properties (name, category).

From the analyst perspective, this gets more complicated. Events are no longer self contained - the information about what *categories* were present is now out-of-band information, that requires co-ordination with the admin to know about. This gets a little messy. For example, if you load the output of events into pandas, what columns you get will now depend on admin parameters. Admins can change this at any time, without a change in schema - causing downstream users to fail. The shape of the data is now different without any such indication to the analyst.

A simplifying suggestion I now have is:

1. Drop categories. Admins need to read the schema anyway, so require them to just explicitly list fields instead. 
2. When emitting events, include all field names regardless of whether they are allowed or not. But for fields that are not allowed, set them to null.

This gives us two things:

1. Admins need to explicitly think about each field they are adding and why
2. All events for a schema will actually *conform* to that schema, but might have missing data. This preserves the shape of the event for all events conforming to a schema, just setting some data as missing instead.

What do you both think? I apologize profusely for providing this feedback so late.

**Comment by yuvipanda** [comment]
I actually take back what I said about categories. For admins, they might not actually know which fields to select! Having a category set by the authors of the schema should definitely help here.

File: jupyter_telemetry/eventlog.py [hunk 7]
```diff
@@ -167,7 +203,25 @@ def record_event(self, schema_name, version, event, timestamp_override=None):
...
+        # allowed categories and properties from the eventlog config.
+        allowed_categories = self.get_allowed_categories(schema_name)
+        allowed_properties = self.get_allowed_properties(schema_name)
+
+        # Iterate through the event properties, and only record the
+        # properties labelled with allowed_categories
+        for property_name, data in event.items():
+            prop_categories = schema["properties"][property_name]["categories"]
+            # If the property is explicitly listed in
+            # the allowed_properties, then include it in the capsule
```
Comment (yuvipanda) [inline jupyter_telemetry/eventlog.py:219]: Instead of excluding it from the capsule, set it to null to indicate missing value?

**Comment by yuvipanda** [comment]
So to recap, the things I think we should do is:

1. Suggest that category names URIs
2. Reserve some for well known categories, like user identifier.
3. When a field shouldn't be recorded, emit it still but set it to null. This preserves shape of data for analysts.

Thank you for coming on this asynchronous journey with me :)

**Comment by yuvipanda** [comment]
Thanks a lot, @Zsailer! This LGTM now, once the test failures are fixed.

**Comment by Zsailer** [comment]
Thanks, @yuvipanda and @blink1073! Merging!

**Pull request merged**